pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+	importerImage       string
+	ovirtPopulatorImage string
+}
+
+// NewForkliftPopulator creates a new instance of the forklift controller.
+func NewForkliftPopulator(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	importerImage string,
+	ovirtPopulatorImage string,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &ForkliftPopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(forkliftPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(forkliftPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			installerLabels: installerLabels,
+		},
+		importerImage:       importerImage,
+		ovirtPopulatorImage: ovirtPopulatorImage,
+	}
+
+	forkliftPopulatorController, err := controller.New(forkliftPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for kind, sourceType := range supportedPopulators {
+		if err := addWatchers(mgr, forkliftPopulatorController, log, kind, sourceType); err != nil {
+			return nil, err
+		}
+	}
+
+	return forkliftPopulatorController, nil
+}
+
+func addWatchers(mgr manager.Manager, c controller.Controller, log logr.Logger, sourceKind string, sourceType client.Object) error {
+	// Setup watches
+	if err := c.Watch(source.Kind(mgr.GetCache(), &corev1.PersistentVolumeClaim{}), handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			pvc := obj.(*corev1.PersistentVolumeClaim)
+			if isPVCForkliftKind(pvc) {
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+
+			// TODO(benny) check dataSourceRef PVC prime
+			if isPVCForkliftKind(pvc) {
+				owner := metav1.GetControllerOf(pvc)
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: owner.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+			return nil
+		}),
+	); err != nil {
+		return err
+	}
+
+	mapDataSourceRefToPVC := func(_ context.Context, obj client.Object) (reqs []reconcile.Request) {
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{
+			dataSourceRefField: getPopulatorIndexKey(apiGroup, sourceKind, obj.GetNamespace(), obj.GetName()),
+		}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return reqs
+	}
+
+	if err := c.Watch(source.Kind(mgr.GetCache(), sourceType),
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Maybe also watch pods that have an ownerRef of a PVC that has the right APIGroup/Kind? That way there is no need to requeue after creating the pod. Once the pod status updates it should trigger a reconcile here.

Reply:
added, although I need to check this some more, not entirely sure I got it right